src/compiler/checker.ts

```diff
@@ -13349,6 +13400,77 @@ namespace ts {
             return links.resolvedType;
         }
 
+        function getTypeFromTemplateTypeNode(node: TemplateTypeNode) {
+            const links = getNodeLinks(node);
+            if (!links.resolvedType) {
+                links.resolvedType = getTemplateType(
+                    [node.head.text, ...map(node.templateSpans, span => span.literal.text)],
+                    map(node.templateSpans, span => span.casing),
+                    map(node.templateSpans, span => getTypeFromTypeNode(span.type)));
+            }
+            return links.resolvedType;
+        }
+
+        function getTemplateType(texts: readonly string[], casings: readonly TemplateCasing[], types: readonly Type[]): Type {

```

Comment:
Yeah, but we're already pretty inconsistent in how we name all of the types related to `TemplateExpression`. `TemplateTypeNode` and `TemplateType` are at least somewhat consistent with `TemplateExpression`. If we are going to rename, it ought to be `TemplateLiteralExpressionNode` and `TemplateLiteralTypeNode` for the AST nodes, and `TemplateLiteralType` for the type. But that would be a breaking change.

Reply:
We could rename and declare some aliases marked `/** @deprecated */` to avoid breaking changes.

If not, I think `TemplateLiteralType` and `TemplateLiteralTypeNode` would help clarify a bit.